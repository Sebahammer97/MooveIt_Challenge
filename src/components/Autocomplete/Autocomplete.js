import React, { Component } from "react";
import PropTypes from "prop-types";
import "./style.css";

class Autocomplete extends Component {
  // Definition of props types to this component
  static propTypes = {
    options: PropTypes.instanceOf(Array).isRequired,
    userInput: PropTypes.string.isRequired,
    onSelect: PropTypes.func,
    onSubmit: PropTypes.func,
  };

  // Assigning default values for the props
  static defaultProps = {
    options: [],
    userInput: "",
  };

  state = {
    activeOption: 0,
    filteredOptions: [],
    showOptions: false,
    userInput: "",
  };

  handleOnChange = (e) => {
    const { options } = this.props;

    const userInput = e.target.value;
    const filteredOptions = options.filter(
      (option) => option.toLowerCase().indexOf(userInput.toLowerCase()) > -1
    );

    this.setState({
      activeOption: 0,
      filteredOptions,
      showOptions: true,
      userInput,
    });
  };

  handleOnClick = (e) => {
    const { onSelect } = this.props;

    this.setState({
      activeOption: 0,
      filteredOptions: [],
      showOptions: false,
      userInput: e.currentTarget.innerText,
    });
    onSelect(e.currentTarget.innerText);
  };

  handleOnKeyDown = (e) => {
    const { onSelect } = this.props;
    const { activeOption, filteredOptions } = this.state;

    if (e.keyCode === 13) {
      // Key Enter
      this.setState({
        activeOption: 0,
        showSuggestions: false,
        filteredOptions,
        showOptions: false,
        userInput: filteredOptions[activeOption],
      });
      onSelect(filteredOptions[activeOption]);
    } else if (e.keyCode === 38) {
      // Key Up
      if (activeOption === 0) {
        return;
      }
      this.setState({ activeOption: activeOption - 1 });
    } else if (e.keyCode === 40) {
      // Key Down
      if (activeOption + 1 === filteredOptions.length) {
        return;
      }
      this.setState({ activeOption: activeOption + 1 });
    }
  };

  render() {
    const { onSubmit } = this.props;
    const {
      handleOnChange,
      handleOnKeyDown,
      handleOnClick,
      state: { activeOption, filteredOptions, showOptions, userInput },
    } = this;

    let optionList;

    if (showOptions && userInput) {
      if (filteredOptions.length) {
        optionList = (
          <ul className="options">
            {filteredOptions.map((optionName, index) => {
              let className;
              if (index === activeOption) {
                className = "option-active";
              }
              return (
                <li
                  className={className}
                  key={optionName}
                  onClick={(e) => handleOnClick(e)}
                >
                  {optionName}
                </li>
              );
            })}
          </ul>
        );
      } else {
        optionList = (
          <div className="no-options">
            <em>No Option!</em>
          </div>
        );
      }
    }

    return (
      <div className="search">
        <input
          type="text"
          className="search-box"
          onChange={(e) => handleOnChange(e)}
          onKeyDown={(e) => handleOnKeyDown(e)}
          value={userInput}
          data-testid="autocomplete"
        />
        <input
          type="submit"
          value=""
          onClick={() => onSubmit(userInput)}
          className="search-btn"
        />
        {optionList}
      </div>
    );
  }
}

export default Autocomplete;
