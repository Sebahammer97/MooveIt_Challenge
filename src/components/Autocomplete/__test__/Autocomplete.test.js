import React, { useState } from "react";
import ReactDOM from "react-dom";
import Autocomplete from "../Autocomplete";

import { render, cleanup, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import renderer from "react-test-renderer";

afterEach(cleanup);

it("autocomplete renders without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(<Autocomplete type={"text"} />, div);
});

function AutocompleteTest() {
  const options = ["Red", "Green", "Blue", "Black", "White", "Yellow"];
  const [input, setInput] = useState("");

  const handleSelect = (value) => {
    setInput(value);
  };

  const handleSubmit = (value) => {
    setInput(value);
  };

  return (
    <Autocomplete
      options={options}
      onSelect={handleSelect}
      onSubmit={handleSubmit}
      userInput={input}
    />
  );
}

it("renders autocomplete input text correctly", () => {
  const utils = render(<AutocompleteTest />);
  const autocomplete = utils.getByTestId("autocomplete");
  fireEvent.change(autocomplete, { target: { value: "abcdefg" } });
  expect(autocomplete.value).toBe("abcdefg");
  fireEvent.change(autocomplete, { target: { value: "123456" } });
  expect(autocomplete.value).toBe("123456");
  fireEvent.change(autocomplete, { target: { value: "a1b2c3d4f5" } });
  expect(autocomplete.value).toBe("a1b2c3d4f5");
});

it("renders autocomplete keydowns", () => {
  const utils = render(<AutocompleteTest />);
  const autocomplete = utils.getByTestId("autocomplete");

  autocomplete.focus();
  fireEvent.change(autocomplete, { target: { value: "b" } });
  fireEvent.keyDown(autocomplete, { keyCode: 40 }); // Key Down
  fireEvent.keyDown(autocomplete, { keyCode: 13 }); // Key Enter

  expect(autocomplete.value).toEqual("Black");
});

it("matches snapshots autocomplete", () => {
  const tree = renderer.create(<Autocomplete />).toJSON();
  expect(tree).toMatchSnapshot();
});
