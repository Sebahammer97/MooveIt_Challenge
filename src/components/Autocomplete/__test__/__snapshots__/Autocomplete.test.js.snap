// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshots autocomplete 1`] = `
<div
  className="search"
>
  <input
    className="search-box"
    data-testid="autocomplete"
    onChange={[Function]}
    onKeyDown={[Function]}
    type="text"
    value=""
  />
  <input
    className="search-btn"
    onClick={[Function]}
    type="submit"
    value=""
  />
</div>
`;
